### PR TITLE
[Perf] Optimize cutlass fp8 scaled mm bypassing padding, 20% kernel performance improvement

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Optimize cutlass fp8 scaled mm bypassing padding

## Test

Using this small script

```bash
import time
import statistics

import torch
from vllm import _custom_ops as ops
from vllm.model_executor.kernels.linear.scaled_mm.cutlass import (
    CutlassFp8BlockScaledMMKernel,
)

torch.cuda.set_device(0)
assert torch.cuda.is_available()

kernel = object.__new__(CutlassFp8BlockScaledMMKernel)
kernel.config = type("Config", (), {"out_dtype": torch.bfloat16})()
kernel.is_hopper = True
kernel.weight_group_shape = [128, 128]


def make_inputs(m, n, k):
    a = torch.randn((m, k), device="cuda", dtype=torch.bfloat16).to(torch.float8_e4m3fn)
    b = torch.randn((n, k), device="cuda", dtype=torch.bfloat16).to(torch.float8_e4m3fn)
    as_ = torch.ones((m, k // 128), device="cuda", dtype=torch.float32)
    bs = torch.ones((n // 128, k // 128), device="cuda", dtype=torch.float32)
    return a, b, as_, bs


def padded(a, b, as_, bs):
    return torch.ops.vllm.padded_cutlass(a, b, as_, bs, [128, 128], torch.bfloat16)


def dispatch(a, b, as_, bs):
    return kernel.apply_block_scaled_mm(a, b, as_, bs)


def bench(fn, args, warmup=50, iters=200, wall_iters=20):
    for _ in range(warmup):
        fn(*args)
    torch.cuda.synchronize()

    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    for _ in range(iters):
        fn(*args)
    end.record()
    torch.cuda.synchronize()
    gpu_us = start.elapsed_time(end) * 1000 / iters

    wall = []
    for _ in range(wall_iters):
        torch.cuda.synchronize()
        t0 = time.perf_counter()
        fn(*args)
        torch.cuda.synchronize()
        wall.append((time.perf_counter() - t0) * 1e6)

    return gpu_us, statistics.median(wall)


print("GPU:", torch.cuda.get_device_name(0))
print("aligned M: compare padded wrapper vs current dispatch")
print("m n k | padded_gpu dispatch_gpu speedup")

for m, n, k in [
    (4, 256, 128),
    (4, 16384, 1024),
    (8, 16384, 1024),
    (16, 16384, 1024),
    (64, 16384, 1024),
    (128, 16384, 1024),
    (256, 16384, 1024),
]:
    args = make_inputs(m, n, k)
    p_gpu, p_wall = bench(padded, args)
    d_gpu, d_wall = bench(dispatch, args)
    print(
        f"{m:2d} {n:5d} {k:4d} | "
        f"{p_gpu:8.2f} {d_gpu:8.2f} {p_gpu/d_gpu:5.2f}x | "
    )
```

And we can get 

```bash
m n k | padded_gpu dispatch_gpu speedup |
 4   256  128 |    21.93    18.33  1.20x | 
 4 16384 1024 |    21.39    17.52  1.22x | 
 8 16384 1024 |    21.12    17.67  1.20x | 
16 16384 1024 |    21.21    17.48  1.21x | 
64 16384 1024 |    21.00    17.49  1.20x | 
128 16384 1024 |    21.16    17.72  1.19x | 
256 16384 1024 |    21.01    17.60  1.19x | 
```

### CODE REVIEW

File: vllm/model_executor/kernels/linear/scaled_mm/cutlass.py [hunk 1]
```diff
@@ -312,6 +312,15 @@ def apply_block_scaled_mm(
...
         if self.is_hopper:
+            # avoid padding when M is already 4-aligned.
+            if A.shape[0] % 4 == 0:
+                return ops.cutlass_scaled_mm(
+                    A,
+                    B.T,
+                    out_dtype=out_dtype,
+                    scale_a=As,
+                    scale_b=Bs.T,
+                )
```
Comment (lgeiger) [inline vllm/model_executor/kernels/linear/scaled_mm/cutlass.py]: We can probably just let this fall through to the else branch when it's already aligned?
```suggestion
        # avoid padding when M is already 4-aligned.
        if self.is_hopper and A.shape[0] % 4 != 0:
```

Comment (yewentao256) [reply vllm/model_executor/kernels/linear/scaled_mm/cutlass.py]: Nice catch, fixed

**Review by Isotr0py — APPROVED** [review]
